builtin/credential/appgroup/path_app.go

```diff
@@ -0,0 +1,931 @@
+	appName := data.Get("app_name").(string)
+	if appName == "" {
+		return logical.ErrorResponse("missing app_name"), nil
+	}
+
+	app, err := b.appEntry(req.Storage, strings.ToLower(appName))
+	if err != nil {
+		return nil, err
+	}
+	if app == nil {
+		return nil, nil
+	}
+
+	// Deleting a field means resetting the value in the entry.
+	app.TokenMaxTTL = (&appStorageEntry{}).TokenMaxTTL
+
+	return nil, b.setAppEntry(req.Storage, appName, app)
+}
+
+func (b *backend) pathAppWrappedUpdate(req *logical.Request, data *framework.FieldData) (*logical.Response, error) {
+	appName := data.Get("app_name").(string)
+	if appName == "" {
+		return logical.ErrorResponse("missing app_name"), nil
+	}
+
+	app, err := b.appEntry(req.Storage, strings.ToLower(appName))
+	if err != nil {
+		return nil, err
+	}
+	if app == nil {
+		return nil, nil
+	}
+
+	if wrappedRaw, ok := data.GetOk("wrapped"); ok {
+		app.Wrapped = time.Duration(wrappedRaw.(int)) * time.Second
+		return nil, b.setAppEntry(req.Storage, appName, app)
+	} else {
+		return logical.ErrorResponse("missing wrapped"), nil
+	}
+}
+
+func (b *backend) pathAppWrappedRead(req *logical.Request, data *framework.FieldData) (*logical.Response, error) {
+	appName := data.Get("app_name").(string)
+	if appName == "" {
+		return logical.ErrorResponse("missing app_name"), nil
+	}
+
+	if app, err := b.appEntry(req.Storage, strings.ToLower(appName)); err != nil {
+		return nil, err
+	} else if app == nil {
+		return nil, nil
+	} else {
+		app.Wrapped = app.Wrapped / time.Second
+		return &logical.Response{
+			Data: map[string]interface{}{
+				"wrapped": app.Wrapped,
+			},
+		}, nil
+	}
+}
+
+func (b *backend) pathAppWrappedDelete(req *logical.Request, data *framework.FieldData) (*logical.Response, error) {
+	appName := data.Get("app_name").(string)
+	if appName == "" {
+		return logical.ErrorResponse("missing app_name"), nil
+	}
+
+	app, err := b.appEntry(req.Storage, strings.ToLower(appName))
+	if err != nil {
+		return nil, err
+	}
+	if app == nil {
+		return nil, nil
+	}
+
+	app.Wrapped = (&appStorageEntry{}).Wrapped
+
+	return nil, b.setAppEntry(req.Storage, appName, app)
+}
+
+func (b *backend) pathAppCredsRead(req *logical.Request, data *framework.FieldData) (*logical.Response, error) {
+	userID, err := uuid.GenerateUUID()
+	if err != nil {
+		return nil, fmt.Errorf("failed to generate UserID:%s", err)
+	}
+	data.Raw["user_id"] = userID
+	return b.handleAppCredsCommon(req, data, false)
+}
+
+func (b *backend) pathAppCredsSpecificUpdate(req *logical.Request, data *framework.FieldData) (*logical.Response, error) {
+	return b.handleAppCredsCommon(req, data, true)
+}
+
+func (b *backend) handleAppCredsCommon(req *logical.Request, data *framework.FieldData, specified bool) (*logical.Response, error) {
+	appName := data.Get("app_name").(string)
+	if appName == "" {
+		return logical.ErrorResponse("missing app_name"), nil
+	}
+
+	userID := data.Get("user_id").(string)
```

Comment:
OK -- I see now why you had the `user_id` in the field data for both -- but I don't really like it because it will show up in `path-help` which Seth assures me people actually use. Here's my suggestion: change handleAppCredsCommon to be `req *logical.Request, data *framework.FieldData, specified string`. Then in `pathAppCredsSpecificUpdate` you can do `return b.handleAppCredsCommon(req, data, data.Get("user_id").(string))`. Then to decide if it's specific or not you simply check whether the string value is `""`.